src/cargo/core/resolver/dep_cache.rs

```diff
@@ -514,25 +516,51 @@ impl RequirementError {
                     .collect();
                 if deps.is_empty() {
                     return match parent {
-                        None => ActivateError::Fatal(anyhow::format_err!(
-                            "Package `{}` does not have the feature `{}`",
-                            summary.package_id(),
-                            feat
-                        )),
+                        None => {
+                            let closest =
+                                closest_msg(&feat.as_str(), summary.features().keys(), |key| &key);
+                            ActivateError::Fatal(anyhow::format_err!(
+                                "Package `{}` does not have the feature `{}`{}",
+                                summary.package_id(),
+                                feat,
+                                closest
+                            ))
+                        }
                         Some(p) => {
                             ActivateError::Conflict(p, ConflictReason::MissingFeatures(feat))
                         }
                     };
                 }
                 if deps.iter().any(|dep| dep.is_optional()) {
                     match parent {
-                        None => ActivateError::Fatal(anyhow::format_err!(
-                            "Package `{}` does not have feature `{}`. It has an optional dependency \
-                             with that name, but that dependency uses the \"dep:\" \
-                             syntax in the features table, so it does not have an implicit feature with that name.",
-                            summary.package_id(),
-                            feat
-                        )),
+                        None => {
+                            let mut features =
+                                features_enabling_dependency_sorted(summary, feat).peekable();
+                            let mut suggestion = String::new();
+                            if features.peek().is_some() {
+                                suggestion = format!(
+                                    "\nDependency `{}` would be enabled by these features:",
+                                    feat
+                                );
+                                for feature in (&mut features).take(3) {
+                                    let _ = write!(&mut suggestion, "\n\t- `{}`", feature);
+                                }
+                                if features.peek().is_some() {
+                                    suggestion.push_str("\n\t  ...");
+                                }
+                            }
+                            ActivateError::Fatal(anyhow::format_err!(
+                                // We left-align this error message here so that rustfmt won't refuse to
+                                // format the remaining parts of this file.
```

Comment:
I can see making this a comment on the PR but leaving this as an artifact in the code adds noise imo

Reply:
Hmm, I would think one would want to know why this formatting is different than elsewhere as a reader of this code.

I can remove if this is blocking, but I don't intend to otherwise, barring some other solution.

Reply:
Speaking for myself in reviewing this, please remove this.

We are not too consistent on our multi-line string formatting generally, a deviation like this is small, this seems like a bug in rustfmt and so would be transient in nature without this having anything tracking it, and has a big comment, raising its visibility more than its relative importance.